---
title: "QC report for single cells ATAC-seq Analysis"
output: 
  html_document:
    front_size: 18px
    highlight: kate
    section_divs: yes
    theme: unite
    toc: yes
    toc_depth: 4
    toc_float: yes
params:
  bc_stat_file: qc_per_bc.bed
  selected_bcs: barcodes.txt
  mapping_qc_file: qc_summary.txt
  fragments_file: fragments.bed
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F)
```

## Pipeline -- scATAC-pro 1.0.0

## MAPPING QC

```{r}
library(data.table)
library(magrittr)
library(kableExtra)
library(RColorBrewer)
mapping_qc = fread(params$mapping_qc_file, header = F)
mapping_qc$frac = round(mapping_qc$V2/mapping_qc$V2[1], 3)
mapping_qc$frac = paste0(100*mapping_qc$frac, '%')

kable(mapping_qc, col.names = NULL, format = 'html') %>%
  kable_styling("striped", full_width = F, position = 'left')

```

## Fragment stats for all barcodes

```{r}
bc_stat = fread(params$bc_stat_file)

frac_peak = sum(bc_stat$total_frags * bc_stat$frac_peak)/sum(bc_stat$total_frags)
frac_mito = sum(bc_stat$total_frags * bc_stat$frac_mito)/sum(bc_stat$total_frags)
frac_promoter = sum(bc_stat$total_frags * bc_stat$frac_promoter)/sum(bc_stat$total_frags)
frac_enh = sum(bc_stat$total_frags * bc_stat$frac_enhancer)/sum(bc_stat$total_frags)
frac_tss = sum(bc_stat$total_frags * bc_stat$frac_tss)/sum(bc_stat$total_frags)

fracs = data.frame(c(frac_peak, frac_mito, frac_promoter, frac_enh, frac_tss))
row.names(fracs) = c('Fraction in peaks', 'Fraction in mitocondrial chromosome',
                    'Fraction in promoters', 'Fraction in Enhancers(ENCODE)', 
                    'Fraction in TSS')
colnames(fracs) = 'pr'
fracs$pr = round(fracs$pr, 3)
fracs$pr = paste0(100*fracs$pr, '%')

kable(fracs, row.names = T, col.names = NULL) %>%
  kable_styling(full_width = F, position = 'left')

```

## Overlapping with functional annotation regions for cell barcode

```{r}
barcodes = fread(params$selected_bcs, header = F)$V1

qc_sele = bc_stat[bc %in% barcodes, ]
qc_nonsele = bc_stat[!bc %in% barcodes, ]


```

```{r fig.width=7, fig.height=5}
## get table or plot as report to html file
set.cols = brewer.pal(n=5, name = 'Dark2')
boxplot(qc_sele$frac_peak, qc_sele$frac_tss, qc_sele$frac_promoter, 
        qc_sele$frac_enh, qc_sele$frac_mito, outline = F,
        ylab = 'frac of reads in annotation region', col = set.cols,
        names = c('Peaks', 'Tss', 'Promoter', 'Enhancer', 'Mito'))

```


```{r fig.width=3, fig.height=4}
set.cols = c('red', 'gray')
boxplot(qc_sele$total_frags,  outline = F, col = set.cols[1],
        ylab = 'Total reads per cell barcode')

boxplot(qc_nonsele$total_frags, outline = F, col = set.cols[2],
        ylab = 'Total reads per non cell barcode')


```

```{r fig.width=6, fig.height=6}
bc_stat[, 'group' := ifelse(bc %in% barcodes, set.cols[1], set.cols[2])]
plot(bc_stat$total_frags, bc_stat$frac_peak, log = 'x', pch = 19, cex = 0.2,
         col = bc_stat$group,
     xlab = 'Total read Pairs', ylab = 'Fraction in Peak')
legend('topright', legend = c('Non-cell', 'cell'), text.col = set.cols[2:1], bty = 'n')

```

## Insersion size Distribution

```{r}
frags = fread(params$fragments_file)
names(frags) = c('chr', 'start', 'end', 'bc', 'ndup')

frags = frags[bc %in% barcodes]
frags[, 'isize' := end - start]

plot(density(frags$isize[frags$isize < 1000], bw = 3),
     lwd = 2, col = brewer.pal(n=3, name = 'Dark2')[1])

```

## Library complexity

```{r}
ndups = sum(frags$ndup) - nrow(frags)
frac.dup = ndups/sum(frags$ndup)

```

